### DISCUSSION TITLE
Correlated sub-queries

### DISCUSSION THREAD

**Discussion by CrossBound** [datalust/seq-tickets#1274] [Ideas]
I realize this will probably not be easy, but Seq really needs a way to do a correlated sub-query. For example, I use the Serilog ASP.NET Core adapter. I need the ability to find any HTTP requests with, say a 401 response, **BUT** do not have a 200 response directly after them. For an additional example: I generally try to use some keyword to tie related logs together, I call it a CorrelationID and it's just a GUID. So if I have several logs over the span of an HTTP request and they all have a common CorrelationID I would like to have the ability to look at one log entry with a particular keyword or error and then correlate it with other log entries with the same correlation ID and only return a list of log entries that match a certain pattern.

As it stands today, there are just some queries that cannot be done in Seq today without writing a special program to do the correlation and searching outside of Seq using the API.

**Comment by nblumhardt** [comment]
Hey @CrossBound - we agree; we've laid some foundations for joins/subqueries in the past year or so, but it will be some time before we're able to bring them forward. Looking forward to it, though!

You may already have seen them, and they probably don't cover all of the cases you're looking for, but `first()`, `last()`, `any()`, and `all()` can achieve a lot.

E.g., just in case it helps in the meantime:

```sql
select any(StatusCode = 200) as ok,
       any(StatusCode = 401) as unauth
from stream
group by CorrelationId
having unaut and not ok
```
